File: backend/cmd/safeguard.go
package cmd

import (
	"fmt"
	"net"
	"net/url"
	"strings"
)

// assertNonProductionURL fails fast if the URL points to a production or public host.
// Seed and simulate commands must only run against local development servers.
func assertNonProductionURL(rawURL string) error {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return fmt.Errorf("invalid URL %q: %w", rawURL, err)
	}
	hostname := strings.ToLower(parsed.Hostname())
	if hostname == "" {
		return fmt.Errorf("URL %q has no hostname", rawURL)
	}
	if isLocalHostname(hostname) {
		return nil
	}
	return fmt.Errorf("refusing to run against %q — seed/simulate commands are for local development only", rawURL)
}

// assertDevOnlyTarget is the guard of the seed and simulate commands: the
// environment must be on the dev-only allow-list and the URL must be local.
// The internal host "server" is local in a deployed compose project too, so
// the URL alone cannot keep staging and production out.
func assertDevOnlyTarget(rawURL, appEnv string) error {
	if !isDevOnlyEnv(appEnv) {
		return fmt.Errorf("refusing to run with APP_ENV=%q — seed/simulate commands are dev-only (allowed: development, test, local, demo, or unset)", appEnv)
	}
	return assertNonProductionURL(rawURL)
}

// isDevOnlyEnv is the allow-list for the dev-only tools. Unset APP_ENV is the
// local-dev default. demo is the public demo environment (ADR 0027), which
// holds synthetic data only; staging, production, and any unknown value stay
// locked out.
func isDevOnlyEnv(appEnv string) bool {
	switch strings.ToLower(strings.TrimSpace(appEnv)) {
	case "", "local", "development", "dev", "test", "demo":
		return true
	default:
		return false
	}
}

func isLocalHostname(hostname string) bool {
	if hostname == "localhost" {
		return true
	}
	if ip := net.ParseIP(hostname); ip != nil && ip.IsLoopback() {
		return true
	}
	return localDockerHostnames[hostname]
}

var localDockerHostnames = map[string]bool{
	"server":                  true,
	"host.docker.internal":    true,
	"gateway.docker.internal": true,
}

File: frontend/e2e/helpers/safeguard.ts
/**
 * Refuses any backend URL that doesn't resolve to a local development host.
 *
 * Mirrors the Go-side guard in `backend/cmd/safeguard.go:assertNonProductionURL`.
 * If `E2E_BACKEND_URL` is ever pointed at staging or production by mistake,
 * the seeder/test scripts can write `E2EPass1234!` into a real environment.
 * Three guard layers — TS (this file), bash (`scripts/lib/assert-local-url.sh`),
 * Go — keep the allowlist consistent and catch each other's bypasses.
 */

const ALLOWED_HOSTNAMES = new Set([
  "localhost",
  "127.0.0.1",
  "::1",
  // Docker-internal hostnames used in development. Keep in sync with
  // backend/cmd/safeguard.go:localDockerHostnames and
  // scripts/lib/assert-local-url.sh.
  "server",
  "server-e2e",
  "host.docker.internal",
  "gateway.docker.internal",
]);

export function assertLocalBackendUrl(rawUrl: string): void {
  let parsed: URL;
  try {
    parsed = new URL(rawUrl);
  } catch (err) {
    throw new Error(
      `invalid backend URL ${JSON.stringify(rawUrl)}: ${
        err instanceof Error ? err.message : String(err)
      }`, { cause: err },
    );
  }

  const host = parsed.hostname.toLowerCase();
  if (!host) {
    throw new Error(`backend URL ${JSON.stringify(rawUrl)} has no hostname`);
  }

  if (ALLOWED_HOSTNAMES.has(host)) return;

  // Any IPv4 loopback (127.0.0.0/8), not just 127.0.0.1.
  if (/^127\.\d+\.\d+\.\d+$/.test(host)) return;

  throw new Error(
    `refusing to run E2E against ${rawUrl} — only localhost, ` +
      `loopback, and Docker-internal hostnames are permitted. ` +
      `Set E2E_BACKEND_URL to a local URL.`,
  );
}
